tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+
+#include <cassert>
+#include <future>
+#include <thread>
+
+
+static std::atomic<int> has_default_objects{0};
+static std::atomic<int> no_default_objects{0};
+static std::atomic<int> no_default_or_assign_objects{0};
+
+
+struct has_default {
+    has_default() : x(0xbad) {
+        ++has_default_objects;
+    }
+    explicit has_default(int n) : x(n) {
+        ++has_default_objects;
+    }
+    has_default(const has_default& v) : x(v.x) {
+        ++has_default_objects;
+    }
+
+    ~has_default() {
+        --has_default_objects;
+    }
+
+    int x;
+};
+
+struct no_default {
+    no_default() = delete;
+    explicit no_default(int n) : x(n) {
+        ++no_default_objects;
+    }
+    no_default(const no_default& v) : x(v.x) {
+        ++no_default_objects;
+    }
+
+    ~no_default() {
+        --no_default_objects;
+    }
+
+    int x;
+};
+
+struct no_default_or_assign {
+    no_default_or_assign() = delete;
+    explicit no_default_or_assign(int n) : x(n) {
+        ++no_default_or_assign_objects;
+    }
+    no_default_or_assign(const no_default_or_assign& v) : x(v.x) {
+        ++no_default_or_assign_objects;
+    }
+
+    void operator=(const no_default_or_assign&) = delete;
+
+    ~no_default_or_assign() {
+        --no_default_or_assign_objects;
+    }
+
+    int x;
+};
+
+template <class F>
+void assert_throws_future_error(F f, std::error_code errc) {
+    try {
+        f();
+    } catch (const std::future_error& e) {
+        assert(e.code() == errc);
+        return;
+    } catch (...) {
+    }
+    assert(false);
+}
+
+template <class T>
+void run_tests() {
+    using promise = std::promise<T>;
+    using future  = std::future<T>;

```

Comment:
Similarly, I recommend choosing different names for these aliases. When I first glanced at this, I thought it was using C++17 Class Template Argument Deduction (where `std::future my_future = my_promise.get_future()` can deduce that you want a `std::future<int>`).

`Promise` and `Future` would be fine. `promise_t`, `future_t`, etc. would work too, or anything else.